### ISSUE TITLE
NullReferenceException when using inheritance, lazy loading, TPH

### ISSUE THREAD

**Issue by Swoffa** [dotnet/efcore#14637]
Describe what is not working as expected.
When using lazy loading, TPH, and having certain relations and inheritance between types (see code below) it is not possible to get a list of one of the inherited types, once added to database. The original code is a much much bigger model migrated from EF6, but for clarity i simplified it heavily, but beeping the error.

If you are seeing an exception, include the full exceptions details (message and stack trace).

```
Exception message: Object reference not set to an instance of an object
Stack trace: at lambda_method(Closure , ValueBuffer )
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalMixedEntityEntry..ctor(IStateManager stateManager, IEntityType entityType, Object entity, ValueBuffer& valueBuffer)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntryFactory.NewInternalEntityEntry(IStateManager stateManager, IEntityType entityType, Object entity, ValueBuffer& valueBuffer)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.StartTrackingFromQuery(IEntityType baseEntityType, Object entity, ValueBuffer& valueBuffer, ISet`1 handledForeignKeys)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityTrackingInfo.StartTracking(IStateManager stateManager, Object entity, ValueBuffer& valueBuffer)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryBuffer.StartTracking(Object entity, EntityTrackingInfo entityTrackingInfo)
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider._TrackEntities[TOut,TIn](IEnumerable`1 results, QueryContext queryContext, IList`1 entityTrackingInfos, IList`1 entityAccessors)+MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at ConsoleApp1.Things.ThingsContext.Test() in C:\Users\Swoffa\source\repos\ConsoleApp1\ConsoleApp1\Things\ThingsContext.cs:line 47
   at ConsoleApp1.Program.Main(String[] args) in C:\Users\Swoffa\source\repos\ConsoleApp1\ConsoleApp1\Program.cs:line 13
```

### Steps to reproduce
Run the `Test` method in the code below to get the Exception.
I was able to "resolve" the problem i 3 different ways, but none of them acceptable:

- Removing `IsRequired` from the relation configuration (commented as solution 1 in code)
- Loading the entities as sub type before accessing the as base type (commented as solution 2 in code)
- Most oddly, renaming `AnotherSpecificSubThing` to `SpecificSubThing2`  (commented as solution 3 in code)

NuGets used:
```
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="2.2.1" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Proxies" Version="2.2.1" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="2.2.1" />
```

```c#
using System.Collections.Generic;
using System.Linq;
using Microsoft.EntityFrameworkCore;

namespace ConsoleApp1.Things
{
    public class ThingsContext : DbContext
    {
        private const string ConnectionString = 
            @"data source=(localdb)\MSSQLLocalDB;initial catalog=Things";

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseLazyLoadingProxies().UseSqlServer(ConnectionString);
            base.OnConfiguring(optionsBuilder);
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<PartialThing>().HasMany(l => l.SpecificSubThings)
                .WithOne(l => l.PartialThing).IsRequired(); //Solution #1: Removing IsRequired
            //modelBuilder.Entity<SpecificSubThing>().HasOne(l => l.PartialThing)
            //    .WithMany(l => l.SpecificSubThings).IsRequired(); //Same error
            modelBuilder.Entity<Thing>();
            modelBuilder.Entity<SpecificSubThing>();
            modelBuilder.Entity<AnotherSpecificSubThing>();
        }

        public static void Test()
        {
            int thingId;
            using (var ctx = new ThingsContext())
            {
                ctx.Database.EnsureDeleted();
                ctx.Database.EnsureCreated();

                var partialThing = new PartialThing();
                var specificSubThing1 = new SpecificSubThing { PartialThing = partialThing };
                var thing = new Thing { SubThings = new List<SubThing> { specificSubThing1 } };
                partialThing.Thing = thing;

                ctx.Set<SubThing>().Add(specificSubThing1);
                //ctx.Set<Thing>().Add(thing); // Gives the Same error
                ctx.SaveChanges();
                thingId = thing.Id;
            }
            using (var ctx = new ThingsContext())
            {
                //ctx.Set<SpecificSubThing>().Where(x => x.Thing.Id == thingId).Load(); //Solution #2: Load as sub type
                var fails = ctx.Set<SubThing>().Where(x => x.Thing.Id == thingId).ToList();
                var alsoFails = ctx.Set<Thing>().Single(x => x.Id == thingId).SubThings;
            }
        }
    }

    public class PartialThing
    {
        public int Id { get; set; }
        public virtual Thing Thing { get; set; }
        public virtual ICollection<SpecificSubThing> SpecificSubThings { get; set; }
    }
    public class Thing
    {
        public int Id { get; set; }
        public virtual ICollection<SubThing> SubThings { get; set; }
    }
    public abstract class SubThing
    {
        public int Id { get; set; }
        public virtual Thing Thing { get; set; }
    }
    public class SpecificSubThing : SubThing
    {
        public virtual PartialThing PartialThing { get; set; }
    }
    public class AnotherSpecificSubThing : SubThing //Solution #3: Renaming to SpecificSubThing2
    {
        public string Name { get; set; }
    }
}
```

### Further technical details
EF Core version: 2.2.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio 2017 15.9.5

**Comment by smitpatel** [comment]
This could be one of those bugs with FK nullability. FK is being marked as required but it is on derived type so properties may ended up being marked required even though nullable in TPH.

Can you try using nightly package as mentioned here https://github.com/aspnet/EntityFrameworkCore/issues/14269#issuecomment-456522029

**Comment by ajcvickers** [comment]
Additional info for triage:
Model:
```
Model: 
  EntityType: AnotherSpecificSubThing Base: SubThing
    Properties: 
      Name (string) 3 3 -1 -1 -1
  EntityType: PartialThing
    Properties: 
      Id (int) Required PK AfterSave:Throw ValueGenerated.OnAdd 0 0 0 -1 0
      ThingId (no field, Nullable<int>) Shadow FK Index 1 1 1 0 1
    Navigations: 
      SpecificSubThings (<SpecificSubThings>k__BackingField, ICollection<SpecificSubThing>) Collection ToDependent SpecificSubThing Inverse: PartialThing PropertyAccessMode.Field 0 -1 2 -1 -1
      Thing (<Thing>k__BackingField, Thing) ToPrincipal Thing PropertyAccessMode.Field 1 -1 3 -1 -1
    Keys: 
      Id PK
    Foreign keys: 
      PartialThing {'ThingId'} -> Thing {'Id'} ToPrincipal: Thing
  EntityType: SpecificSubThing Base: SubThing
    Properties: 
      PartialThingId (no field, int) Shadow Required FK Index 3 3 3 2 2
    Navigations: 
      PartialThing (<PartialThing>k__BackingField, PartialThing) ToPrincipal PartialThing Inverse: SpecificSubThings PropertyAccessMode.Field 1 -1 4 -1 -1
    Foreign keys: 
      SpecificSubThing {'PartialThingId'} -> PartialThing {'Id'} ToDependent: SpecificSubThings ToPrincipal: PartialThing
  EntityType: SubThing Abstract
    Properties: 
      Id (int) Required PK AfterSave:Throw ValueGenerated.OnAdd 0 0 0 -1 0
      Discriminator (no field, string) Shadow Required AfterSave:Throw 1 1 -1 0 -1
      ThingId (no field, Nullable<int>) Shadow FK Index 2 2 1 1 1
    Navigations: 
      Thing (<Thing>k__BackingField, Thing) ToPrincipal Thing Inverse: SubThings PropertyAccessMode.Field 0 -1 2 -1 -1
    Keys: 
      Id PK
    Foreign keys: 
      SubThing {'ThingId'} -> Thing {'Id'} ToDependent: SubThings ToPrincipal: Thing
  EntityType: Thing
    Properties: 
      Id (int) Required PK AfterSave:Throw ValueGenerated.OnAdd 0 0 0 -1 0
    Navigations: 
      SubThings (<SubThings>k__BackingField, ICollection<SubThing>) Collection ToDependent SubThing Inverse: Thing PropertyAccessMode.Field 0 -1 1 -1 -1
    Keys: 
      Id PK
```
Generated tables:
```sql
      CREATE TABLE [Thing] (
          [Id] int NOT NULL IDENTITY,
          CONSTRAINT [PK_Thing] PRIMARY KEY ([Id])
      );
      CREATE TABLE [PartialThing] (
          [Id] int NOT NULL IDENTITY,
          [ThingId] int NULL,
          CONSTRAINT [PK_PartialThing] PRIMARY KEY ([Id]),
          CONSTRAINT [FK_PartialThing_Thing_ThingId] FOREIGN KEY ([ThingId]) REFERENCES [Thing] ([Id]) ON DELETE NO ACTION
      );
      CREATE TABLE [SubThing] (
          [Id] int NOT NULL IDENTITY,
          [ThingId] int NULL,
          [Discriminator] nvarchar(max) NOT NULL,
          [Name] nvarchar(max) NULL,
          [PartialThingId] int NULL,
          CONSTRAINT [PK_SubThing] PRIMARY KEY ([Id]),
          CONSTRAINT [FK_SubThing_PartialThing_PartialThingId] FOREIGN KEY ([PartialThingId]) REFERENCES [PartialThing] ([Id]) ON DELETE CASCADE,
          CONSTRAINT [FK_SubThing_Thing_ThingId] FOREIGN KEY ([ThingId]) REFERENCES [Thing] ([Id]) ON DELETE NO ACTION
      );
```
Query logs:
```	  
dbug: Microsoft.EntityFrameworkCore.Query[10104]
      => Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor
      Optimized query model:
      'from SubThing x in DbSet<SubThing>
      where Property([x], "ThingId") == (Nullable<int>)__thingId_0
      select [x]'
dbug: Microsoft.EntityFrameworkCore.Query[10107]
      => Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor
      (QueryContext queryContext) => IEnumerable<SubThing> _InterceptExceptions(
      |__ source: IEnumerable<SubThing> _TrackEntities(
      |   |__ results: IEnumerable<SubThing> _ShapedQuery(
      |   |   |__ queryContext: queryContext,
      |   |   |__ shaperCommandContext: SelectExpression:
      |   |   |       SELECT [x].[Id], [x].[Discriminator], [x].[ThingId], [x].[Name], [x].[PartialThingId]
      |   |   |       FROM [SubThing] AS [x]
      |   |   |       WHERE [x].[Discriminator] IN (N'SpecificSubThing', N'AnotherSpecificSubThing') AND ([x].[ThingId] = @__thingId_0),
      |   |   |__ shaper: BufferedEntityShaper<SubThing>),
      |   |__ queryContext: queryContext,
      |   |__ entityTrackingInfos: { itemType: SubThing },
      |   |__ entityAccessors: List<Func<SubThing, object>>
      |       {
      |           Func<SubThing, SubThing>,
      |       }),
      |__ contextType: ConsoleApp1.Things.ThingsContext,
      |__ logger: DiagnosticsLogger<Query>,
      |__ queryContext: queryContext)
dbug: Microsoft.EntityFrameworkCore.Database.Command[20100]
      Executing DbCommand [Parameters=[@__thingId_0='?' (DbType = Int32)], CommandType='Text', CommandTimeout='30']
      SELECT [x].[Id], [x].[Discriminator], [x].[ThingId], [x].[Name], [x].[PartialThingId]
      FROM [SubThing] AS [x]
      WHERE [x].[Discriminator] IN (N'SpecificSubThing', N'AnotherSpecificSubThing') AND ([x].[ThingId] = @__thingId_0)
```

**Comment by smitpatel** [comment]
> EntityType: SpecificSubThing Base: SubThing
    Properties: 
      PartialThingId (no field, int) Shadow Required FK Index 3 3 3 2 2

Seems bit out of place. Either property should be marked not required. (though table does mark it as null) or query pipeline is interpreting data incorrectly.

**Comment by Swoffa** [comment]
Problem still exist with
```
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="2.2.2-servicing-35839" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Proxies" Version="2.2.2-servicing-35839" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="2.2.2-servicing-35839" />
```
...and
```
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.0.0-preview.19074.3" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Proxies" Version="3.0.0-preview.19074.3" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.0.0-preview.19074.3" />
```

**Comment by smitpatel** [comment]
@Swoffa - Thanks for verifying. Marking this as bug.

**Comment by Swoffa** [comment]
Thanks!
For now, I got rid of the error and achieved what I wanted (querying works and Exception when attempting to save SpecificSubThing without a PartialThing). The "solution" was to explicitly add a non-nullable foreign key:
```
    public class SpecificSubThing : SubThing
    {
        public int PartialThingId { get; set; }
        public virtual PartialThing PartialThing { get; set; }
    }
```

...and removing `IsRequired()` from `OnModelCreating`:
```
            modelBuilder.Entity<PartialThing>().HasMany(l => l.SpecificSubThings)
                .WithOne(l => l.PartialThing);
```

But I think there are some weird things going on causing the original error, and the `IsRequired()` is not producing the expected output, so probably worth investigating further.

**Comment by AndriySvyryd** [comment]
Fixed in 432a4a18ede80610fa05e70d3d4afc5e129e8ef7
